test/make_dist.py

```diff
@@ -52,115 +51,78 @@ def build_dist():
-                        sys.stderr.write(f"\r{ total_size // MB } MB")
-
-                    tario.write(block)
-
-                # clear the download progress in tty mode
-                if total_size is not None:
-                    sys.stderr.write("\r                             \r")
-
-                break
-
-        except urllib.error.HTTPError as e:
+            subprocess.check_call(["git", "-C", dist_git_checkout, "fetch", "origin", "sha-" + sha, "--depth=1"])
+            break
+        except subprocess.CalledProcessError:
             retries -= 1
 
             if retries == 0:
-                message(f"make_dist: Downloading {download_url} failed:", e)
-                return None
+                message(f"make_dist: Downloading pre-built dist for SHA {sha} failed")
+                return False
 
-            message(f"make_dist: {download_url} not yet available, waiting...")
+            message(f"make_dist: pre-built dist for {sha} not yet available, waiting...")
             time.sleep(30)
 
-    tario.seek(0)
-    with tarfile.open(fileobj=tario) as ftar:
-        names = ftar.getnames()
-        try:
-            names.remove('.')
-        except ValueError:
-            pass
-        if len(names) != 1 or not names[0].endswith(".tar.xz"):
-            message("make_dist: expected tar with exactly one tar.xz member")
-            return None
-        ftar.extract(names[0])
-        tar_path = os.path.realpath(names[0])
-
-    # Extract npm/webpack related files locally for speeding up the build and allowing integration tests to run
-    unpack_paths = [d for d in ["dist", "node_modules", "package-lock.json", "tools/debian/copyright"] if not os.path.exists(d)]
-    if unpack_paths:
-        message("make_dist: Extracting from tarball:", ' '.join(unpack_paths))
-        prefix = os.path.basename(tar_path).split('.tar')[0] + '/'
-        prefixed_unpack_paths = [prefix + d for d in unpack_paths]
-        subprocess.check_call(["tar", "--touch", "--strip-components=1", "-xf", tar_path] + prefixed_unpack_paths)
-        # after the above, dist manifests may be newer than copyright due to tar touching order
-        # restore file time stamps as they would be after a build
-        subprocess.check_call(["touch", "tools/debian/copyright"])
-
-    return tar_path
+    # built files need to have current time, not commit time, to satisfy make dependencies
+    tar_cmd = { False: ["tar", "-x"], True: ["tar", "--touch", "-x"] }
+    for (unpack_path, touch) in [("node_modules", False),
+                                 ("package-lock.json", False),
+                                 ("dist", True),
+                                 ("tools/debian/copyright", True)]:
+        if os.path.exists(unpack_path):
+            continue
+        message(f"make_dist: Extracting cached {unpack_path}...")
+        p_git = subprocess.Popen(["git", "-C", dist_git_checkout, "archive", "FETCH_HEAD", unpack_path],
```

Comment:
`FETCH_HEAD` is unsafe here: you might be building more than one thing on the same system, sharing the cache directory.  Use the tag.  Also, more `--git-dir`.

Reply:
I had some trouble with the tag, as `git fetch origin <tagname>` does not quite fetch the actual tag, you get "not a valid object name: sha-ba9...". Indeed refs/tags/ is empty. And `git fetch --tag origin <tagname>` fetches *all* tags. Your PR #15700 has exactly the same problem. I'll do some googling/experimentation.

Reply:
I replaced the fetching with the magic incantation to *actually* fetch the tag, now it works fine. I also followed up to your PR to do the same there.